src/app/clusters/client-monitoring-server/client-monitoring-server.cpp

```diff
@@ -42,51 +50,163 @@ class ClientMonitoringAttributeAccess : public app::AttributeAccessInterface
+
+private:
+    CHIP_ERROR ReadExpectedClients(EndpointId endpoint, AttributeValueEncoder & encoder);
+};
+
+CHIP_ERROR ClientMonitoringAttributeAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == ClientMonitoring::Id);
+
+    switch (aPath.mAttributeId)
     {
-        VerifyOrDie(aPath.mClusterId == ClientMonitoring::Id);
+    case ClientMonitoring::Attributes::ExpectedClients::Id:
+        return ReadExpectedClients(aPath.mEndpointId, aEncoder);
 
-        switch (aPath.mAttributeId)
-        {
-        case ClientMonitoring::Attributes::ExpectedClients::Id:
-            // TODO : Implement Client monitoring registration table
-            return CHIP_IM_GLOBAL_STATUS(UnsupportedRead);
+    default:
+        break;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ClientMonitoringAttributeAccess::Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder)
+{
+    return CHIP_IM_GLOBAL_STATUS(UnsupportedWrite);
+}
+
+CHIP_ERROR ClientMonitoringAttributeAccess::ReadExpectedClients(EndpointId endpoint, AttributeValueEncoder & encoder)
+{
+    FabricIndex fabric = encoder.AccessingFabricIndex();
 
-        default:
-            break;
+    return encoder.EncodeList([fabric](const auto & encoder) -> CHIP_ERROR {
+        ClientMonitoringRegistrationTable clientMonitoringRegistrationTable(chip::Server::GetInstance().GetPersistentStorage());
+        CHIP_ERROR err = clientMonitoringRegistrationTable.LoadFromStorage(fabric);
+
+        if (err == CHIP_NO_ERROR)
+        {
+            ReturnErrorOnFailure(encoder.Encode(clientMonitoringRegistrationTable.GetClientRegistrationEntry()));
+        }
+        else if (err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND)
+        {
+            // No Entries saved for this fabric. Return empty list
+            err = CHIP_NO_ERROR;
         }
 
-        return CHIP_NO_ERROR;
-    }
+        return err;
+    });
+}
+
+ClientMonitoringAttributeAccess gAttribute;
+
+} // namespace
+
+/**********************************************************
+ * ClientMonitoringServer Implementation
+ *********************************************************/
+
+InteractionModel::Status ClientMonitoringServer::StayAwakeRequestCommand(const ConcreteCommandPath & commandPath)
+{
+    // TODO: Implementent stay awake logic for end device
+    return InteractionModel::Status::UnsupportedCommand;
+}
+
+InteractionModel::Status
+ClientMonitoringServer::RegisterClientMonitoringCommand(CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
+                                                        const Commands::RegisterClientMonitoring::DecodableType & commandData)
+{
+    FabricIndex fabric = commandObj->GetAccessingFabricIndex();
+
+    ClientMonitoringRegistrationTable table(chip::Server::GetInstance().GetPersistentStorage());
```

Comment:
It's too bad that this relies on chip::Server::instance, but I guess it can be fixed later.

Reply:
Created an issue to revisit this.
#24289